// sentiment/lib/index.js
var tokenize = require('./tokenize');
var languageProcessor = require('./language-processor');

/**
 * Constructor
 * @param {Object} options - Instance options
 */
var Sentiment = function (options) {
    this.options = options;
};

/**
 * Registers the specified language
 *
 * @param {String} languageCode
 *     - Two-digit code for the language to register
 * @param {Object} language
 *     - The language module to register
 */
Sentiment.prototype.registerLanguage = function (languageCode, language) {
    languageProcessor.addLanguage(languageCode, language);
};

/**
 * Performs sentiment analysis on the provided input 'phrase'.
 *
 * @param {String} phrase
 *     - Input phrase
 * @param {Object} opts
 *     - Options
 * @param {Object} opts.language
 *     - Input language code (2 digit code), defaults to 'en'
 * @param {Object} opts.extras
 *     - Optional sentiment additions to AFINN (hash k/v pairs)
 * @param {function} callback
 *     - Optional callback
 * @return {Object}
 */
Sentiment.prototype.analyze = function (phrase, opts, callback) {
    // Parse arguments
    if (typeof phrase === 'undefined') phrase = '';
    if (typeof opts === 'function') {
        callback = opts;
        opts = {};
    }
    opts = opts || {};

    var languageCode = opts.language || 'ch';
    var labels = languageProcessor.getLabels(languageCode);

    // Merge extra labels
    if (typeof opts.extras === 'object') {
        labels = Object.assign(labels, opts.extras);
    }

    // Storage objects
    var tokens      = tokenize(phrase),
        score       = 0,
        // 乐(PA,PE) 好(PD,PH,PG,PB,PK) 怒(NA) 哀(NB,NJ,NH,PF) 
        // 惧(NI,NC,NG) 恶(NE,ND,NN,NK,NL) 惊(PC)
        moods        = {
            "happy": 0,
            "good": 0,
            "anger": 0,
            "grief": 0,
            "fear": 0,
            "bad": 0,
            "surprise": 0
        },
        words       = [],
        positive    = [],
        negative    = [],
        calculation = [];

    // Iterate over tokens
    var i = tokens.length;
    while (i--) {
        var obj = tokens[i];
        if (!labels.hasOwnProperty(obj)) continue;
        words.push(obj);
        // console.log(score, typeof score)
        // Apply scoring strategy
        var tokenMood = {
            side: labels[obj].side,
            mood: labels[obj].mood
        }
        // eslint-disable-next-line max-len
        tokenMood.side = languageProcessor.applyScoringStrategy(languageCode, tokens, i, tokenMood.side);
        tokenMood = languageProcessor.distinguishScoringStrategy(languageCode, score, tokens[i], tokenMood);
        if (tokenMood.side > 0) positive.push(obj);
        if (tokenMood.side < 0) negative.push(obj);
        score += tokenMood.side;
        if(tokenMood.mood == "PA" || tokenMood.mood == "PE"){
            moods.happy += 1;
        } else if(tokenMood.mood == "PD" || tokenMood.mood == "PH" || tokenMood.mood == "PG" || tokenMood.mood == "PB" || tokenMood.mood == "PK"){
            moods.good += 1;
        } else if(tokenMood.mood == "NA"){
            moods.anger += 1;
        } else if(tokenMood.mood == "NB" || tokenMood.mood == "NJ" || tokenMood.mood == "NH" || tokenMood.mood == "PF"){
            moods.grief += 1;
        } else if(tokenMood.mood == "NI" || tokenMood.mood == "NC" || tokenMood.mood == "NG"){
            moods.fear += 1;
        } else if(tokenMood.mood == "NE" || tokenMood.mood == "ND" || tokenMood.mood == "NN" || tokenMood.mood == "NK" || tokenMood.mood == "NL"){
            moods.bad += 1;
        } else if(tokenMood.mood == "PC"){
            moods.surprise += 1;
        }
        
        var zipObj = {}; 
        // Calculations breakdown
        zipObj[obj] = tokenMood;
        calculation.push(zipObj);
    }

    var result = {
        score:          score,
        comparative:    tokens.length > 0 ? score / tokens.length : 0,
        moods:          moods,
        calculation:    calculation,
        tokens:         tokens,
        words:          words,
        positive:       positive,
        negative:       negative
    };

    // Handle optional async interface
    if (typeof callback === 'function') {
        process.nextTick(function () {
            callback(null, result);
        });
    } else {
        return result;
    }
};

module.exports = Sentiment;
